lib/sql.js

```diff
@@ -156,200 +228,892 @@ SqlConnector.prototype.tableEscaped = function (model) {
     return;
   }
-  var self = this;
-  var idQuery = this.idColumnEscaped(model) + ' = ' + this._escapeIdValue(model, id);
-  var sql = 'SELECT * FROM ' +
-    this.tableEscaped(model) + ' WHERE ' + idQuery + ' LIMIT 1';
+  var idName = this.idName(model);
+  var where = {};
+  where[idName] = id;
+  this.destroyAll(model, where, options, cb);
+};
+// Alias to `destroy`. Juggler checks `destroy` only.
+SqlConnector.prototype.delete = SqlConnector.prototype.destroy;
+
+/**
+ * Build the `DELETE FROM` SQL statement
+ * @param {String} model The model name
+ * @param {Object} where The where object
+ * @param {Object} options Options object
+ * @returns {ParameterizedSQL} The SQL DELETE FROM statement
+ */
+SqlConnector.prototype.buildDelete = function(model, where, options) {
+  var deleteStmt = new ParameterizedSQL('DELETE FROM ' +
+    this.tableEscaped(model));
+  deleteStmt.merge(this.buildWhere(model, where));
+  return this.parameterize(deleteStmt);
+};
 
-  this.query(sql, function (err, data) {
-    var result =  (data && data.length >= 1) ? data[0] : null;
-    if (callback) callback(err, self.fromDatabase(model, result));
+/**
+ * Delete all matching model instances
+ *
+ * @param {String} model The model name
+ * @param {Object} where The where object
+ * @param {Object} options The options object
+ * @param {Function} cb The callback function
+ */
+SqlConnector.prototype.destroyAll = function(model, where, options, cb) {
+  var self = this;
+  var stmt = this.buildDelete(model, where, options);
+  this.execute(stmt.sql, stmt.params, options, function(err, info) {
+    var affectedRows = self.getCountForAffectedRows(model, info);
+    if (cb) {
+      cb(err, {count: affectedRows});
+    }
   });
 };
+// Alias to `destroyAll`. Juggler checks `destroyAll` only.
+SqlConnector.prototype.deleteAll = SqlConnector.prototype.destroyAll;
 
 /**
- * Delete a model instance by id value
+ * ATM, this method is not used by loopback-datasource-juggler dao, which
+ * maps `updateAttributes` to `update` with a `where` filter that includes the
+ * `id` instead.
+ *
+ * Update attributes for a given model instance
  * @param {String} model The model name
  * @param {*} id The id value
- * @param {Function} callback The callback function
+ * @param {Object} data The model data instance containing all properties to
+ * be updated
+ * @param {Object} options Options object
+ * @param {Function} cb The callback function
+ * @private
  */
-SqlConnector.prototype.delete =
-SqlConnector.prototype.destroy = function destroy(model, id, callback) {
-  if (!isIdValuePresent(id, callback, true)) {
+SqlConnector.prototype.updateAttributes = function(model, id, data, options, cb) {
+  if (!isIdValuePresent(id, cb)) {
     return;
   }
-  var sql = 'DELETE FROM ' + this.tableEscaped(model) + ' WHERE ' +
-    this.idColumnEscaped(model) + ' = ' + this._escapeIdValue(model, id);
+  var idName = this.idName(model);
+  delete data[idName];
```

Comment:
FWIW, I believe it's the job of DAO to remove `idName` property from the data sent to updateAttributes. I suppose it does not make any harm to remove it here too, although I don't think all connectors do the same.